text/3416-feature-metadata.md

```diff
@@ -0,0 +1,401 @@
+attribute in Rust source. The value can be a boolean, string, or an object with
+`since` or `note` keys. Schema rules are as follows:
+
+-   If a boolean value, `false` indicates not deprecated and `true` indicates
+    deprecated
+-   If an object, the keys `since` and/or `note` can be specified
+-   An empty object is not allowed to avoid the ambiguous
+    `foo = { deprecated = {} }`
+-   A string `foo = { deprecated = "my msg" }` will be equivalent to if that string
+    was specified in the `note` field:
+    `foo = { deprecated = { note = "my msg" } }`
+-   If not specified, the default is `false`
+
+If a downstream crate attempts to use a feature marked `deprecated`, Cargo
+should produce a warning that contains the `note`. This warning should not be
+emitted for crates that reexport the feature under a feature also marked
+deprecated. For example: crate `foo` exports feature `phooey`, and crate `bar`
+exports feature `barred = ["foo/phooey"]`. If `foo` markes `bar` as deprecated,
+running any cargo action on `bar` will emit a warning unless `barred` is also
+marked `deprecated.
+
+Accessing this information will require access to the manifest as it will not be
+in the index.
+
+### A note on `since`
+
+The exact behavior of the `since` key is not provided in this RFC as there are
+decisions related to resolution that need to be made. The generally accepted
+concept is that there should be a warning if a deprecated feature is used _and_
+there is something actionable to resolve this issue for all downstream crates -
+but the details of how best to do this are not yet clear. Please see [discussion
+on since].
+
+If the exact behavior of `since` does not reach consensus before `deprecated` is
+nearing stabilization, this key can stabilized separately or dropped entirely.
+
+## `public`
+
+`public` is a boolean value that defaults to `true`. It can be thought of as
+`pub` in Rust source files, with the exception of being true by default. If set
+to `false`, Cargo should forbid its use with an error message on any downstream
+crates.
+
+The default `true` is not consistent with [`public_private_dependencies`] or
+Rust's `pub`, but is a reasonable default to be consistent with the current
+behavior so that either `feature = []` or `feature = { "enables" = [] }` will
+result in the same configuration.
+
+The name `public` was chosen in favor of `pub` to be consistent with the
+[`public_private_dependencies`] RFC, and to match the existing style of using
+non-truncated words as keys.
+
+In general, marking a feature `public = false` should make tooling treat the
+feature as non-public API. This is described as the following:
+
+-   The feature is always usable within the same crate:
+    -   Enablement by other features, e.g. `foo = { enables =
+        [some-private-feature] }`, is allowed
+    -   Using the feature in integration tests is allowed
+    -   Using the feature in benchmarks is allowed
```

Comment:
I assume what you mean is that `#[cfg(feature = ..)]` works in the source and the user can specify the features on the command line?

Otherwise, there really isn't a way for a test or benchmark to "use" a feature

Reply:
Moved to #3487